backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt

```diff
@@ -160,29 +162,26 @@ internal class ObjCExportCodeGenerator(
 
     private val objCTypeAdapters = mutableListOf<ObjCTypeAdapter>()
 
-    internal fun generate(
-            generatedClasses: Collection<ClassDescriptor>,
-            categoryMembers: Map<ClassDescriptor, List<CallableMemberDescriptor>>,
-            topLevel: Map<SourceFile, List<CallableMemberDescriptor>>
-    ) {
-        generatedClasses.forEach {
-            objCTypeAdapters += createTypeAdapter(it, categoryMembers[it].orEmpty())
-
-            if (!it.isInterface) {
-                val className = namer.getClassOrProtocolName(it).binaryName
-                val superClass = it.getSuperClassOrAny()
-                val superClassName = namer.getClassOrProtocolName(superClass).binaryName
-
-                dataGenerator.emitEmptyClass(className, superClassName)
-                // Note: it is generated only to be visible for linker.
-                // Methods will be added at runtime.
+    internal fun generate(spec: ObjCExportCodeSpec) {
+        spec.types.forEach {
+            objCTypeAdapters += when (it) {
+                is ObjCClassForKotlinClass -> {
+                    val superClass = it.superClassNotAny ?: objCClassForAny
+
+                    dataGenerator.emitEmptyClass(it.binaryName, superClass.binaryName)
+                    // Note: it is generated only to be visible for linker.
+                    // Methods will be added at runtime.
+
+                    createTypeAdapter(it, superClass)
+                }
+
+                is ObjCProtocolForKotlinInterface -> createTypeAdapter(it, superClass = null)
```

Comment:
Shalln't we have catching `else` here?

Reply:
This `when` is used as an expression and thus forced to be exhaustive.